assets/js/modules/analytics-4/datastore/containers.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * `modules/analytics-4` data store: containers.
+ *
+ * Site Kit by Google, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import Data from 'googlesitekit-data';
+import { MODULES_ANALYTICS_4 } from './constants';
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+
+const fetchGetGoogleTagContainerStore = createFetchStore( {
+	baseName: 'getGoogleTagContainer',
+	controlCallback( { measurementID } ) {
+		return API.get(
+			'modules',
+			'analytics-4',
+			'container-lookup',
+			{ destinationID: measurementID },
+			{
+				useCache: false,
+			}
+		);
+	},
+	reducerCallback( state, container, { measurementID } ) {
+		return {
+			...state,
+			containers: {
+				...state.containers,
+				[ measurementID ]: container,
+			},
+		};
+	},
+	argsToParams( measurementID ) {
+		return { measurementID };
+	},
+	validateParams( { measurementID } = {} ) {
+		invariant( measurementID, 'measurementID is required.' );
+	},
+} );
+
+const fetchGetGoogleTagContainerDestinationsStore = createFetchStore( {
+	baseName: 'getGoogleTagContainerDestinations',
+	controlCallback( { gtmAccountID, gtmContainerID } ) {
+		return API.get(
+			'modules',
+			'analytics-4',
+			'container-destinations',
+			{ accountID: gtmAccountID, internalContainerID: gtmContainerID },
+			{
+				useCache: false,
+			}
+		);
+	},
+	reducerCallback(
```

Comment:
This looks like a good opportunity to use [`createReducer`](https://github.com/google/site-kit-wp/blob/c4dd8ea26fb0b522c885e769c974c76aee26a0ef/assets/js/googlesitekit/data/create-reducer.js#L32) to create an Immer-enabled reducer callback, in order to simplify the reducer here.

It would look something like this. Still not super elegant, but easier to make sense of:

```js
	reducerCallback: createReducer(
		( state, containerDestinations, { gtmAccountID, gtmContainerID } ) => {
			state.containerDestinations[ gtmAccountID ] =
				state.containerDestinations[ gtmAccountID ] || {};

			state.containerDestinations[ gtmAccountID ][ gtmContainerID ] =
				state.containerDestinations[ gtmAccountID ][ gtmContainerID ] ||
				[];

			state.containerDestinations[ gtmAccountID ][ gtmContainerID ].push(
				...containerDestinations
			);
		}
	),
```

Reply:
Thanks for the great suggestion, @techanvil. I didn't think of it.

>It would look something like this. Still not super elegant, but easier to make sense of:

The above code look good 👍